proposals/unconstrained-type-parameter-annotation.md

```diff
@@ -0,0 +1,114 @@
+# Unconstrained type parameter annotation: `T??`
+
+## Summary
+
+Support `??` as a nullable annotation for type parameter references when the type parameters are not constrained
+to a value type or reference type.
+
+## Design
+
+### Syntax
+The parser is extended to accept `??` in addition to `?` for nullable types.
+`??` is a single token with no characters separating the two `?` characters.
+```
+nullable_type
+    : type '?'
+    | type '??'
+    ;
+```
+
+`T?` is not allowed as the type in an `is` expression and so neither is `T??`.
+```C#
+_ = a is T?? b;     // (a is T) ?? b
+_ = a is T?? b : c; // syntax error
+```
+
+`??` is not allowed for `class` constraint.
+```C#
+interface I<T>
+    where T : class?? // syntax error
+{
+}
+```
+
+`T?` is not allowed for a type parameter `T` that is not constrained to a value or reference type.
+```C#
+T? F1<T>() { ... }                   // error: T must be value type or non-nullable reference type
+T? F2<T>() where T : notnull { ... } // error: T must be value type or non-nullable reference type
+```
+
+`T??` is not allowed for a type parameter `T` that is constrained to a value or reference type.
+```C#
+T?? F3<T>() where T : class { ... }  // error: T is a reference type
+T?? F4<T>() where T : struct { ... } // error: T is a value type
+```
+
+`??` is not allowed for annotating types other than type parameters.
+```C#
+int?? F5() { ... }  // error: int is not a type parameter
+T[]?? F6<T> { ... } // error: T[] is not a type parameter
+```
+
+_Should `??` be allowed or required for type parameters with inherited constraints that constrain the type parameter sufficiently?_
+```C#
+abstract class A<T>
+{
+    internal abstract U?? F<U>() where U : T;
+}
+class B1 : A<string>
+{
+    internal override U?? F<U>() => default; // Is ?? allowed or required?
+}
+class B2 : A<int>
+{
+    internal override U?? F<U>() => default; // Is ?? allowed or required?
+}
+class B3 : A<int?>
+{
+    internal override U?? F<U>() => default; // Is ?? allowed or required?
+}
+```
+
+`T??` may be used interchangably with `[MaybeNull]T` in cases where `T` is not constrained to a value type or reference type.
```

Comment:
I really hope this lands as `T?` at the end - for the same reasons that the team were avoiding introducing `T!`.

Furthermore, the fact that you need to change annotations `T??` -> `T?` if you add a constraint is a pain point.

Reply:
I really hope unconstrained `T?` will sooner or later end up as an unconstrained `T?` that is `T?` for reference types or `T?` for value types.

Reply:
>I really hope unconstrained T? will sooner or later end up as an unconstrained T? that is T? for reference types or T? for value types.

That would probably require runtime changes - personally I'm ok with `[MaybeNull]` for the time being. `T??` would be just an oddball that I don't want to have to deal with.